### PR TITLE
adds R.stableSort and tests

### PR DESCRIPTION
**Pull request by ConorLinehan**
Addresses #1135.

### CODE REVIEW

File: src/stableSort.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *      R.sort(diff, [4,2,7,5]); //=> [2, 4, 5, 7]
+ */
+module.exports = _curry2(function stableSort(comparator, list) {
+  var copyList = _slice(list);
+
+  return copyList.sort(function(a, b) {
+    if (comparator(a, b) === 0) {
+      return -(copyList.indexOf(a) - copyList.indexOf(b));
+    } else {
+      return comparator(a, b);
```
Comment (buzzdecafe) [inline src/stableSort.js]: why call `comparator` twice?

**Comment by davidchambers** [comment]
Should we update `R.sort` rather than define a new function?

File: src/stableSort.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *      var diff = function(a, b) { return a - b; };
+ *      R.sort(diff, [4,2,7,5]); //=> [2, 4, 5, 7]
+ */
+module.exports = _curry2(function stableSort(comparator, list) {
+  var copyList = _slice(list);
+
+  return copyList.sort(function(a, b) {
+    if (comparator(a, b) === 0) {
+      return -(copyList.indexOf(a) - copyList.indexOf(b));
```
Comment (scott-christopher) [inline src/stableSort.js]: An alternative to scanning the list twice with `indexOf` here would be to zip up the input list with the index of each field and then use the index to perform the secondary comparison.

e.g. something like:
```js
var stableSort = R.curry(function (comparator, list) {
  var withIndex = R.zip(list, R.range(0, list.length));
  var sortedWithIndex = withIndex.sort(function(a, b) {
    var order = comparator(a[0], b[0]);
    return order === 0 ? a[1] - b[1] : order;
  });
  return R.map(R.head, sortedWithIndex);
});
```

I'd be interested to see how these two compare for different sized lists.

**Comment by CrossEye** [comment]
> Should we update `R.sort` rather than define a new function?

:+1:

**Comment by CrossEye** [comment]
> An alternative to scanning the list twice with `indexOf` here would be to zip up the input list with the index of each field and then use the index to perform the secondary comparison. [ ... ]

> I'd be interested to see how these two compare for different sized lists.

Agreed.  I've used that before for stable sorts.

**Comment by ConorLinehan** [comment]
@scott-christopher  Heres a benchmark of the different methods http://jsperf.com/stable-sort-ramda

**Comment by oskarkv** [comment]
Hey guys, where are you on this?

**Comment by CrossEye** [comment]
@oskarkook:

Nowhere at the moment.  It's fallen between the cracks.

@ConorLinehan: Any interest in addressing the concerns brought up in comments here?  I would like to include this.

**Comment by ConorLinehan** [comment]
@CrossEye Yup will do :)

**Comment by ConorLinehan** [comment]
> Should we update R.sort rather than define a new function?

After doing the benchmarks (http://jsperf.com/stable-sort-ramda) with both discussed methods do we still want to update `R.sort` or create the new `R.stableSort` function with the following implementation

```
   var stableSort1 = R.curry(function(comparator, list) {
      var withIndex = R.zip(list, R.range(0, list.length));
      var sortedWithIndex = withIndex.sort(function(a, b) {
        var order = comparator(a[0], b[0]);
        return order === 0 ? a[1] - b[1] : order;
      });
      return R.map(R.head, sortedWithIndex);
    });
```

**Comment by CrossEye** [comment]
I would look for ways to optimize the solution, but at a guess, we may need to have a separate `stableSort` version.

**Comment by scott-christopher** [comment]
I've added an extra benchmark case [here] which uses the same method, but hand-rolls the zip-with-index, etc. It shows a slight boost of performance in Chrome, but negligible in FF. :shrug:

```js
    var stableSort3 = R.curry(function(comparator, list) {
      var sorted = [];
      var idx;
    
      idx = 0;
      while (idx < list.length) {
        sorted.push([list[idx], idx]);
        idx += 1;
      }
    
      sorted.sort(function(a, b) {
        var order = comparator(a[0], b[0]);
        return order === 0 ? a[1] - b[1] : order;
      });
    
      idx = 0;
      while (idx < list.length) {
        sorted[idx] = sorted[idx][0];
        idx += 1;
      }
      return sorted;
    });
```

[here]: http://jsperf.com/stable-sort-ramda/2

File: src/stableSort.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Returns a copy of the list, sorted according to the comparator function, which should accept two values at a
+ * time and return a negative number if the first value is smaller, a positive number if it's larger, and zero
+ * if they are equal, values of the same key retain their original order. Please note that this is a **copy** of the list.
+ *  It does not modify the original.
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a,a -> Number) -> [a] -> [a]
+ * @param {Function} comparator A sorting function :: a -> b -> Int
```
Comment (davidchambers) [inline src/stableSort.js]: Let's remove `a -> b -> Int` here. It's confusing that it doesn't match `(a,a -> Number)` above, and at any rate it's unnecessary to duplicate the information.

File: test/stableSort.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    var sortByLength = R.sort(function(a, b) {return a.length - b.length;});
+    eq(sortByLength(['one', 'two', 'three', 'four', 'five', 'six']),
+       ['one', 'two', 'six', 'four', 'five', 'three']);
+  });
+
+  it('maintains stability', function() {
+    var list = [
+      {type: 'Apple', price: 10},
+      {type: 'Pear', price: 100},
+      {type: 'Bannana', price: 10},
```
Comment (davidchambers) [inline test/stableSort.js]: `s/Bannana/Banana/`

**Comment by davidchambers** [comment]
> I would look for ways to optimize the solution, but at a guess, we may need to have a separate `stableSort` version.

I don't like the sound of this. Users would then need to choose between the "fast" sorting function which doesn't always do the right thing and the "slow" sorting function which does. This is incidental complexity. I'm imagining the entries in @svozza's [_I want to…_][1] table. _I want to sort a list quickly_ and _I want to **stably** sort a list (and I'm not in a hurry)_. ;)


[1]: https://github.com/ramda/ramda/wiki/What-Function-Should-I-Use%3F

**Comment by CrossEye** [comment]
> I don't like the sound of this. Users would then need to choose between the "fast" sorting function which doesn't always do the right thing and the "slow" sorting function which does. 

Which could also be written as, _I want to sort a list_  and _I want to sort a list so that equivalent items retain their original order_, without mentioning speed or with just trailing parenthetical _(faster)_ and _(slower)_ notation.

We can certainly look for more performant sorters.  There is no reason to rely on `Array.prototype.sort`.  If we want to take the complexity hit, we can use [Timsort][ts], or we could try a mergesort.  There are plenty of options.

Or we can live with the performance hit.  That might well be the best, but it certainly would seem worth pursuing the options before making that decision.


  [ts]: https://en.wikipedia.org/wiki/Timsort

**Comment by davidchambers** [comment]
My view is that we should make a decision on behalf of our users. If we consider stable sorting to be important enough to justify a significant performance hit, we should update `R.sort`. If not, we should include a recipe for stable sorting in the cookbook and reference it in the description of `R.sort`.

**Comment by CrossEye** [comment]
I can probably agree to that.  It's probably not worth having a second function.  But I think there is more investigation necessary to find if we can have our cake and still eat at least some of it.

**Comment by ConorLinehan** [comment]
Paraphrasing from this, http://stackoverflow.com/questions/3026281/array-sort-sorting-stability-in-different-browsers, it seems chrome uses a none stable sort and are sticking to it(https://bugs.chromium.org/p/v8/issues/detail?id=90)
Last post
>Project Member Comment 58 by jkummerow@chromium.org, Nov 10, 2015
Using O(n) stack space is out of the question, as the stack is way too limited (less than 1 MB). Observe how the existing implementation limits itself to O(log n) stack space even in the worst case (i.e. pathologically unlucky pivot elements) by avoiding the recursive descent into the larger half of the array (lines 1036-1042).

All the other browsers seem to use a stable sort. So from the above benchmarks [here](http://jsperf.com/stable-sort-ramda/2) we would only be able to get near to firefox. Currently we're roughly 50% slower than firefox 47, So I suppose it's picking an acceptable tolerance?

**Comment by CrossEye** [comment]
But see also a comparison with various merge-sort implementation:

  http://jsperf.com/stable-sort-comparison/5

Timsort is quite a bit faster, but it also is substantially more complex to implement.  But most merge-sort implementation are stable, I believe.  That might be enough.

If we want stability, we obviously cannot use the native sort.  Even if all the current crop of browsers did use a stable version, there is simply no guarantee.  We could build atop it as above, or we could use our own, mergesort, timsort, or one of many other choices.

**Comment by CrossEye** [comment]
So, should we consider an implementation like this instead?:


```js
module.exports = _curry2(function sort(comparator, list) {

  return msort(_slice(list), 0, list.length);

  function msort(list, begin, end) {
    var size = end - begin;
    if (size < 2) return;
    var middle = begin + Math.floor(size / 2);

    msort(list, begin, middle);
    msort(list, middle, end);
    merge(list, begin, middle, end);
    return list;
  }

  function merge(list, begin, middle, end) {
    while (begin < middle) {
      if (comparator(list[begin], list[middle]) > 0) {
        var val = list[begin];
        list[begin] = list[middle];
        insert(list, middle, end, val);
      }
      begin += 1;
    }
  }

  function insert(list, begin, end, val) {
    while (begin + 1 < end && comparator(list[begin + 1], val) < 0) {
      swap(list, begin, begin + 1);
      begin += 1;
    }
    list[begin] = val;
  }

  function swap(list, a, b) {
    var tmp = list[a];
    list[a] = list[b];
    list[b] = tmp;
  }
});
```

(Perhaps this would need to be pulled into an _internal version to be reused by `sortBy`.)

Of course this depends hugely on internal mutation, but that doesn't escape the function, so I don't see any issues.

This is a really basic merge sort and maybe one of the ones at http://jsperf.com/stable-sort-comparison/5 would be faster, but anything based on the same ideas might not only add stability but also actually be more performant (for now) than `Array.prototype.sort`.

**Comment by ajhyndman** [comment]
I'd be keen to see this move ahead.  I think a not-guaranteed-to-be-stable sort implementation is a bit of a foot-gun in JavaScript, and I think Ramda would do well to guard users from that trap.

**Comment by davidchambers** [comment]
It's worth mentioning that [`Z.sort`][1] performs a stable sort. Ramda would get stable sorting as part of the package if #2347 is merged one day.


[1]: https://github.com/sanctuary-js/sanctuary-type-classes#sort

**Comment by GingerPlusPlus** [comment]
https://github.com/ramda/ramda/issues/2926#issuecomment-564125453

**Pull request closed without merging**